File: py/PointsInACircle.py
import math

from Image import Image
from Color import Color

import Utilities
from RNG31Transform.Uniform import Uniform
from RNG31Transform.Normal import Normal

WIDTH    = 1024
HEIGHT   =  620
MARGIN   =   10
BOX_SIZE =  (WIDTH - (3.0 * MARGIN)) / 2.0

RECT_UNIFORM_X = MARGIN
RECT_UNIFORM_Y = (HEIGHT - BOX_SIZE) / 2
RECT_UNIFORM_W = BOX_SIZE
RECT_UNIFORM_H = BOX_SIZE

RECT_NORMAL_X = MARGIN + MARGIN + BOX_SIZE
RECT_NORMAL_Y = (HEIGHT - BOX_SIZE) / 2
RECT_NORMAL_W = BOX_SIZE
RECT_NORMAL_H = BOX_SIZE

WIDTH    = 1024
HEIGHT   =  620
MARGIN   =   10
DIAMETER =  (WIDTH - (3.0 * MARGIN)) / 2.0

RADIUS = DIAMETER / 2.0
CENTER_UNCORRECTED_X = MARGIN + RADIUS
CENTER_UNCORRECTED_Y = HEIGHT / 2.0
CENTER_CORRECTED_X = MARGIN + MARGIN + DIAMETER + RADIUS
CENTER_CORRECTED_Y = CENTER_UNCORRECTED_Y


if __name__ == "__main__":
    img = Image(WIDTH, HEIGHT)
    img.fill(Color.BLK)

    arng = Utilities.getRNGCore()
    urng = Uniform(arng)
    nrng = Normal(arng)

    ITERATIONS = 250000
    for _ in range(ITERATIONS):
        radius = urng.nextFloat(RADIUS)
        radians = urng.nextFloat(2.0 * math.pi)
        x = radius * math.cos(radians) + CENTER_UNCORRECTED_X
        y = radius * math.sin(radians) + CENTER_UNCORRECTED_Y
        Utilities.addSamplePoint(img, int(x), int(y))

    for _ in range(ITERATIONS):
        x, y = urng.pointInCircle(RADIUS)
        x += CENTER_CORRECTED_X
        y += CENTER_CORRECTED_Y
        Utilities.addSamplePoint(img, int(x), int(y))

    Utilities.writeBMPImage(img, "PointsInACircle.bmp")
    print("- The more times a point is sampled, the brighter it gets.");
    print("- The circle on the left contains random points with an uncorrected Uniform sampling.");
    print("- The circle on the right contains random points with a corrected Uniform sampling.");

    exit(0)

File: cpp/PointsInACircle.cpp
#include <cstdio>
#include <cmath>

#include "Utilities.h"
#include "RNG31Core/AbstractRNGCore.h"
#include "RNG31Transform/Uniform.h"

#include "Image.h"
#include "Color.h"

// Image Properties
#define WIDTH    1280
#define HEIGHT    768
#define MARGIN     10
#define DIAMETER  625

const double RADIUS = DIAMETER / 2.0;
const double CENTER_UNCORRECTED_X = MARGIN + RADIUS;
const double CENTER_UNCORRECTED_Y = HEIGHT / 2;
const double CENTER_CORRECTED_X = MARGIN + MARGIN + DIAMETER + RADIUS;
const double CENTER_CORRECTED_Y = CENTER_UNCORRECTED_Y;

int main(int argc, char **argv)
{
    RNG31::AbstractRNGCore *arng = getRNGCore();
    RNG31::Uniform urng(arng);

    Image img(WIDTH, HEIGHT);
    img.fill(Color::BLK);

    const int ITERATIONS = 450000;
    for(int index = 0; index < ITERATIONS; ++index) {
        double radius = urng.nextFloat(RADIUS);
        double radians = urng.nextFloat(2.0 * M_PI);
        double x = radius * cos(radians) + CENTER_UNCORRECTED_X;
        double y = radius * sin(radians) + CENTER_UNCORRECTED_Y;
        addSamplePoint(&img, (int)x, (int)y);
    }

    for(int index = 0; index < ITERATIONS; ++index) {
        double x, y;
        urng.pointInCircle(RADIUS, x, y);
        x += CENTER_CORRECTED_X;
        y += CENTER_CORRECTED_Y;
        addSamplePoint(&img, (int)x, (int)y);
    }

    writeBMPImage(&img, (char*)"PointsInACircle.bmp");
    printf("- The more times a point is sampled, the brighter it gets.\n");
    printf("- The circle on the left contains random points with an uncorrected Uniform sampling.\n");
    printf("- The circle on the right contains random points with a corrected Uniform sampling.\n");

    delete arng;
    return 0;
}
